### PR TITLE
EL-4043-Toolbar-Search

### PR DESCRIPTION
**Pull request by cmallon367**
#### Checklist

* [x] The contents of this PR are mine to submit. No third party code or other material is being committed.
* [x] New third party dependencies (if any) are linked via `package.json`. Third party dependencies are licensed under one of the following: Apache, MIT, BSD, Mozilla Public License, Eclipse Public License, or Oracle Binary Code License.
* [x] The code in this PR does not contain export-restricted encryption algorithms.

#### Ticket / Issue

https://portal.digitalsafe.net/browse/EL-4043 

#### Description of Proposed Changes

Updated the screenshots for the toolbar search 
Added an e2e to verify that the typeahead appears correctly when opened.
Reduced padding right to fix a white flicker that was happening on expanding the toolbar search

#### Documentation CI URL

https://pages.github.houston.softwaregrp.net/sepg-docs-qa/UXAspects_CI_UXAspects_EL-4043-Toolbar-Search/

#### Downstream Build(s)

https://jenkins.swinfra.net/job/SEPG/job/caf/view/Developer/job/caf~ux-aspects-micro-focus~EL-4043-Toolbar-Search~CI/

### CODE REVIEW

File: e2e/pages/app/toolbar-search/toolbar-search.testpage.component.html [hunk 1]
```diff
@@ -1,83 +1,98 @@
...
 
-</div>
+    <p class="m-t-md">Searched for:
+        <strong id="searchedFor">{{searchedFor}}</strong>
+    </p>
 
-<p class="m-t-md">Searched for:
-    <strong id="searchedFor">{{searchedFor}}</strong>
-</p>
+    <div id="mastheadSearchText" class="m-t-md">
```
Comment (ashley-hunter) [inline e2e/pages/app/toolbar-search/toolbar-search.testpage.component.html:94]: I don't think this is required, this is testing the selection functionality of a typeahead, which should be covered by the typeahead tests.

File: e2e/pages/app/toolbar-search/toolbar-search.testpage.component.ts [hunk 2]
```diff
@@ -30,4 +31,53 @@ export class ToolbarSearchTestPageComponent {
...
+
+    mastheadSearchExpanded: boolean;
+    mastheadSearchText: string;
+    mastheadSearchedFor: string = '';
+    values: ReadonlyArray<string> = [];
+    searchDropdownOpen: boolean;
+
+    input: string = '';
+
+    loadOptionsFn = this.loadOptions.bind(this);
```
Comment (ashley-hunter) [inline e2e/pages/app/toolbar-search/toolbar-search.testpage.component.ts:50]: Can we get rid of the load function and just use an array of a few static values

File: e2e/pages/app/toolbar-search/toolbar-search.testpage.component.ts [hunk 3]
```diff
@@ -30,4 +31,53 @@ export class ToolbarSearchTestPageComponent {
...
...erbia', 'Seychelles', 'Sierra Leone', 'Singapore', 'Sint Maarten (Dutch part)', 'Slovakia', 'Slovenia', 'Solomon Islands', 'Somalia', 'South Africa', 'South Georgia and The South Sandwich Islands', 'South Sudan', 'Spain', 'Sri Lanka', 'Sudan', 'Suriname', 'Svalbard and Jan Mayen', 'Swaziland', 'Sweden', 'Switzerland', 'Syrian Arab Republic', 'Taiwan, Province of China', 'Tajikistan', 'Tanzania, United Republic of', 'Thailand', 'Timor-leste', 'Togo', 'Tokelau', 'Tonga', 'Trinidad and Tobago', 'Tunisia', 'Turkey', 'Turkmenistan', 'Turks and Caicos Islands', 'Tuvalu', 'Uganda', 'Ukraine', 'United Arab Emirates', 'United States Minor Outlying Islands', 'Uruguay', 'Uzbekistan', 'Vanuatu', 'Venezuela, Bolivarian Republic of', 'Viet Nam', 'Virgin Islands, British', 'Virgin Islands, U.S.', 'Wallis and Futuna', 'Western Sahara', 'Yemen', 'Zambia', 'Zimbabwe', 'MMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMM', 'Daenerys of the House Targaryen, the First of Her Name, The Unburnt, Queen of the Andals, the Rhoynar and the First Men, Queen of Meereen, Khaleesi of the Great Grass Sea, Protector of the Realm, Lady Regent of the Seven Kingdoms, Breaker of Chains and Mother of Dragons'];
```
Comment (ashley-hunter) [inline e2e/pages/app/toolbar-search/toolbar-search.testpage.component.ts:79]: We don't need all this dummy data, a simple list of ['One', 'Two', 'Three'] would suffice

File: e2e/tests/components/toolbar-search/toolbar-search.e2e-spec.ts [hunk 4]
```diff
@@ -31,6 +31,21 @@ describe('Toolbar Search', () => {
...
+        // selecting country with clicking
+        await page.clickOnDropdown();
+        expect(await page.confirmDropdownIsExpanded()).toBeFalsy();
+
+        expect(await imageCompare('toolbar-search-typeahead-open')).toEqual(0);
+
+        // Enter search text
+        await page.leftInput.sendKeys('United States');
+
+        expect(await page.getSelectedLocationText()).toBe('United States');
```
Comment (ashley-hunter) [inline e2e/tests/components/toolbar-search/toolbar-search.e2e-spec.ts:45]: I think this can be removed, we don't need to test selection functionality as this would be covered by typeahead tests

File: e2e/tests/components/toolbar-search/toolbar-search.e2e-spec.ts [hunk 5]
```diff
@@ -103,7 +118,7 @@ describe('Toolbar Search', () => {
         await page.leftButton.click();
 
         // Verify that the states are unchanged
-        expect(await page.leftInput.isDisplayed()).toBeTruthy();
+        expect(await page.leftInput.isDisplayed()).toBeFalsy();
```
Comment (ashley-hunter) [inline e2e/tests/components/toolbar-search/toolbar-search.e2e-spec.ts:121]: Why was this change required?

File: e2e/pages/app/toolbar-search/toolbar-search.testpage.component.html [hunk 6]
```diff
@@ -1,83 +1,93 @@
...
-                #searchField
+        <div class="demo-toolbar-left">
+            <ux-toolbar-search id="searchLeft"
+                [alwaysExpanded]="alwaysExpanded"
+                [(expanded)]="expanded"
+                direction="right"
+                background="white"
+                (search)="onSearch($event)">
+
+            <input
```
Comment (ashley-hunter) [inline e2e/pages/app/toolbar-search/toolbar-search.testpage.component.html:11]: The indentation here is incorrect

File: e2e/pages/app/toolbar-search/toolbar-search.testpage.component.html [hunk 7]
```diff
@@ -1,83 +1,93 @@
...
+            <ux-typeahead
+                #searchTypeahead
+                class="typeahead-example"
+                [(open)]="searchDropdownOpen"
+                [filter]="mastheadSearchText"
+                [options]="options"
+                (optionSelected)="input = $event.option">
+            </ux-typeahead>
+
+            <button *ngIf="mastheadSearchText" type="button"
```
Comment (ashley-hunter) [inline e2e/pages/app/toolbar-search/toolbar-search.testpage.component.html:33]: For some reason this button and the search button are showing up as blue?

File: e2e/pages/app/toolbar-search/toolbar-search.testpage.component.ts [hunk 8]
```diff
@@ -1,5 +1,6 @@
-import { ChangeDetectionStrategy, Component } from '@angular/core';
+import { ChangeDetectionStrategy, Component, ViewChild, ElementRef } from '@angular/core';
 import { FormControl, FormGroup } from '@angular/forms';
+import { TypeaheadKeyService } from '@ux-aspects/ux-aspects';
```
Comment (ashley-hunter) [inline e2e/pages/app/toolbar-search/toolbar-search.testpage.component.ts:3]: This import is unused

File: e2e/pages/app/toolbar-search/toolbar-search.testpage.component.ts [hunk 9]
```diff
@@ -30,4 +31,38 @@ export class ToolbarSearchTestPageComponent {
         // Execute search here
         this.searchedFor = this.form.controls.search.value;
     }
+
+    @ViewChild('searchField')
```
Comment (ashley-hunter) [inline e2e/pages/app/toolbar-search/toolbar-search.testpage.component.ts:35]: A lot of the variables and functions in this class are no used anywhere and should be removed. Also class properties should be above any methods. I also notice a few methods are missing return types.

File: e2e/tests/components/toolbar-search/toolbar-search.po.spec.ts [hunk 10]
```diff
@@ -20,7 +20,22 @@ export class ToolbarSearchPage {
...
+    async clickOnDropdown() {
+        await this.leftButton.click();
+        return this.leftInput.click();
+    }
+
+    confirmDropdownIsExpanded() {
+        return this.left.$('ux-typeahead-all-options').isPresent();
+    }
+
+    getSelectedLocationText() {
```
Comment (ashley-hunter) [inline e2e/tests/components/toolbar-search/toolbar-search.po.spec.ts:37]: I don't think this is needed any more?

File: e2e/pages/app/toolbar-search/toolbar-search.testpage.component.ts [hunk 11]
```diff
@@ -14,20 +14,37 @@ export class ToolbarSearchTestPageComponent {
     searchedFor: string = '';
     alwaysExpanded: boolean = false;
 
+    toolbarExpanded: boolean;
```
Comment (ashley-hunter) [inline e2e/pages/app/toolbar-search/toolbar-search.testpage.component.ts:17]: I dont think the `searchText` variable above is used any more

Comment (ashley-hunter) [inline e2e/pages/app/toolbar-search/toolbar-search.testpage.component.ts:17]: I don't think `toolbarExpanded` is every used in the HTML

File: e2e/pages/app/toolbar-search/toolbar-search.testpage.component.html [hunk 12]
```diff
@@ -1,83 +1,96 @@
...
+                    (blur)="searchDropdownOpen = false"
+                    (click)="searchDropdownOpen = true"/>
+
+                <ux-typeahead
+                    #searchTypeahead
+                    class="typeahead-example"
+                    [(open)]="searchDropdownOpen"
+                    [filter]="mastheadSearchText"
+                    [options]="options"
+                    (optionSelected)="input = $event.option">
```
Comment (ashley-hunter) [inline e2e/pages/app/toolbar-search/toolbar-search.testpage.component.html:30]: I think this should be:

```html
(optionSelected)="mastheadSearchText = $event.option"
```
You can then remove the `input` property from the class

File: e2e/pages/app/toolbar-search/toolbar-search.testpage.component.ts (hunk 11 above)
Comment (cmallon367) [reply e2e/pages/app/toolbar-search/toolbar-search.testpage.component.ts:17]: toolbarExpanded is used below in the onToolbarSearch

Comment (ashley-hunter) [reply e2e/pages/app/toolbar-search/toolbar-search.testpage.component.ts:17]: Yes, but its not used in the HTML so it isn't actually doing anything?

Comment (cmallon367) [reply e2e/pages/app/toolbar-search/toolbar-search.testpage.component.ts:17]: searchText is being used for the onSearch below and that's being used on toolbar search (search)="onSearch($event)

Comment (ashley-hunter) [reply e2e/pages/app/toolbar-search/toolbar-search.testpage.component.ts:17]: The searchText in that function is being passed in, it is not using `this.searchText` which is the class property.

image

Also the whole `onToolbarSearch` is not called from the HTML so it too can be removed?

File: e2e/tests/components/toolbar-search/toolbar-search.e2e-spec.ts [hunk 13]
```diff
@@ -31,6 +31,14 @@ describe('Toolbar Search', () => {
 
     });
 
+    it('should display the name of the selected number', async () => {
```
Comment (DeanHaire) [inline e2e/tests/components/toolbar-search/toolbar-search.e2e-spec.ts:34]: I don't think this name makes it clear what the test is doing. Is it just checking that 'One' is selected by default?

File: e2e/tests/components/toolbar-search/toolbar-search.e2e-spec.ts [hunk 14]
```diff
@@ -31,6 +31,14 @@ describe('Toolbar Search', () => {
...
+
+        await page.clickOnDropdown();
+        expect(await page.confirmDropdownIsExpanded()).toBeFalsy();
```
Comment (DeanHaire) [inline e2e/tests/components/toolbar-search/toolbar-search.e2e-spec.ts:37]: Should this not be true? The dropdown looks open in the screenshot? Or maybe I'm misunderstanding the purpose of this test

File: e2e/tests/components/toolbar-search/toolbar-search.po.spec.ts [hunk 15]
```diff
@@ -23,4 +23,14 @@ export class ToolbarSearchPage {
 
     // button to set always expanded state
     alwaysExpandedBtn = element(by.id('always-expanded-btn'));
+
+    async clickOnDropdown() {
```
Comment (DeanHaire) [inline e2e/tests/components/toolbar-search/toolbar-search.po.spec.ts:27]: Would `openDropdown` be a better name for this? `clickOnDropdown` might imply just clicking somewhere within the open dropdown.

File: e2e/tests/components/toolbar-search/toolbar-search.po.spec.ts [hunk 16]
```diff
@@ -23,4 +23,14 @@ export class ToolbarSearchPage {
...
+        await this.leftButton.click();
+        return this.leftInput.click();
+    }
+
+    confirmDropdownIsExpanded() {
```
Comment (DeanHaire) [inline e2e/tests/components/toolbar-search/toolbar-search.po.spec.ts:32]: `isDropdownExpanded` might be a simpler name for this?

File: e2e/tests/components/toolbar-search/toolbar-search.po.spec.ts [hunk 17]
```diff
@@ -23,4 +23,5 @@ export class ToolbarSearchPage {
 
     // button to set always expanded state
     alwaysExpandedBtn = element(by.id('always-expanded-btn'));
+
```
Comment (donna-marijne) [inline e2e/tests/components/toolbar-search/toolbar-search.po.spec.ts:26]: Remove this change.

File: src/components/toolbar-search/toolbar-search.component.less [hunk 18]
```diff
@@ -30,7 +30,7 @@ ux-toolbar-search {
         }
 
         [uxToolbarSearchField] {
-            padding-right: 35px;
+            padding-right: 20px;
```
Comment (donna-marijne) [inline src/components/toolbar-search/toolbar-search.component.less:33]: Update the pull request description to state why this change is needed please.

File: e2e/pages/app/toolbar-search/toolbar-search-typeahead/toolbar-search-typeahead.testpage.component.html [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+            <button type="button" class="btn btn-link btn-icon button-secondary">
+                <ux-icon name="add"></ux-icon>
+            </button>
+
+            <button type="button" class="btn btn-link btn-icon button-secondary">
+                <ux-icon name="edit"></ux-icon>
+            </button>
+        </div>
+
+    </div>
```
Comment (donna-marijne) [inline e2e/pages/app/toolbar-search/toolbar-search-typeahead/toolbar-search-typeahead.testpage.component.html:58]: Indentation is still off.

File: e2e/pages/app/toolbar-search/toolbar-search-typeahead/toolbar-search-typeahead.testpage.component.ts [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+    styleUrls: ['./toolbar-search-typeahead.testpage.component.less'],
+    changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class ToolbarSearchTypeaheadTestPageComponent {
+
+    expanded: boolean;
+    searchedFor: string = '';
+    alwaysExpanded: boolean = false;
+
+    mastheadSearchText: string;
```
Comment (donna-marijne) [inline e2e/pages/app/toolbar-search/toolbar-search-typeahead/toolbar-search-typeahead.testpage.component.ts:16]: There's no masthead in this example so this variable name doesn't make sense.

File: e2e/tests/components/toolbar-search/toolbar-search-typeahead/toolbar-search-typeahead.e2e-spec.ts [hunk 21]
```diff
@@ -0,0 +1,25 @@
+import { protractor } from 'protractor';
+import { imageCompare } from '../../common/image-compare';
+import { ToolbarSearchTypeheadPage } from './toolbar-search-typeahead.po.spec';
+
+
+const ec = protractor.ExpectedConditions;
```
Comment (donna-marijne) [inline e2e/tests/components/toolbar-search/toolbar-search-typeahead/toolbar-search-typeahead.e2e-spec.ts:6]: This does not appear to be used.

File: e2e/tests/components/toolbar-search/toolbar-search-typeahead/toolbar-search-typeahead.po.spec.ts [hunk 22]
```diff
@@ -0,0 +1,27 @@
+import { browser, by, element } from 'protractor';
+
+export class ToolbarSearchTypeheadPage {
+
+    async getPage(): Promise<void> {
+        await browser.get('#/toolbar-search/toolbar-search-typeahead');
+    }
```
Comment (donna-marijne) [inline e2e/tests/components/toolbar-search/toolbar-search-typeahead/toolbar-search-typeahead.po.spec.ts:7]: Functions belong underneath the property declarations.

File: e2e/tests/components/toolbar-search/toolbar-search-typeahead/toolbar-search-typeahead.po.spec.ts [hunk 23]
```diff
@@ -0,0 +1,27 @@
...
+    // ux-toolbar-search component
+    search = element(by.id('search'));
+    toolbarInput = this.search.$('[uxToolbarSearchField]');
+    searchButton = this.search.$('[uxToolbarSearchButton]');
+    searchClear = this.search.$('.ux-toolbar-search-clear');
+
+    // Most recently submitted search query
+    searchedFor = element(by.id('searchedFor'));
+
+    async openDropdown() {
```
Comment (donna-marijne) [inline e2e/tests/components/toolbar-search/toolbar-search-typeahead/toolbar-search-typeahead.po.spec.ts:18]: Add the return type to this function.

File: e2e/tests/components/toolbar-search/toolbar-search-typeahead/toolbar-search-typeahead.po.spec.ts [hunk 24]
```diff
@@ -0,0 +1,27 @@
...
+        await this.searchButton.click();
+        return this.toolbarInput.click();
+    }
+
+    isDropdownExpanded() {
```
Comment (donna-marijne) [inline e2e/tests/components/toolbar-search/toolbar-search-typeahead/toolbar-search-typeahead.po.spec.ts:23]: Add the return type to this function.

File: src/components/toolbar-search/toolbar-search.component.less (hunk 18 above)
Comment (donna-marijne) [inline src/components/toolbar-search/toolbar-search.component.less:33]: The text now overlaps the icon so please revert this change and find an alternative solution.
image

**Review by donna-marijne — CHANGES_REQUESTED** [review]
Thanks, but the padding change is not appropriate, see the screenshot.

**Review by donna-marijne — APPROVED** [review]
Changes look good, thank you.

**Review by DeanHaire — APPROVED** [review]

**Review by ashley-hunter — APPROVED** [review]
All good, thanks
